dipy/denoise/gibbs.py

```diff
@@ -0,0 +1,314 @@
+    cosk1 = 1.0 + np.cos(K1)
+    G1[1:-1, 1:-1] = cosk0 / (cosk0+cosk1)
+    G0[1:-1, 1:-1] = cosk1 / (cosk0+cosk1)
+
+    # Boundaries
+    G1[1:-1, 0] = G1[1:-1, -1] = 1
+    G1[0, 0] = G1[-1, -1] = G1[0, -1] = G1[-1, 0] = 1/2
+    G0[0, 1:-1] = G0[-1, 1:-1] = 1
+    G0[0, 0] = G0[-1, -1] = G0[0, -1] = G0[-1, 0] = 1/2
+
+    return G0, G1
+
+
+def _gibbs_removal_2d(image, n_points=3, G0=None, G1=None):
+    """ Decreases gibbs ringing of a 2D image.
+
+    Parameters
+    ----------
+    image : 2D ndarray
+        Matrix cotaining the 2D image.
+    n_points : int, optional
+        Number of neighbour points to access local TV (see note). Default is
+        set to 3.
+    G0 : 2D ndarray, optional.
+        Weights for the image corrected along axis 1. If not given, the
+        function estimates them using function :func:`_weights`
+    G1 : 2D ndarray
+        Weights for the image corrected along axis 1. If not given, the
+        function estimates them using function :func:`_weights`
+
+    Returns
+    -------
+    imagec : 2D ndarray
+        Matrix with gibbs oscilantions reduced along axis a.
+    tv : 2D ndarray
+        Global TV which show variation not removed by the algorithm (edges,
+        anatomical variation, non-oscilatory component of gibbs artefact
+        normally present in image background, etc.)
+
+    Note
+    ----
+    This function decreases the effects of gibbs oscilations based on the
+    analysis of local total variation (TV) along the two axis of the image.
+    Although artefact correction is done based on each point primary adjanced
+    neighbors, total variation should be accessed in a larger range of
+    neigbors. If you want to adjust the number of the neigbors to be
+    considered in TV calculation please change parameter nn.
+
+    References
+    ----------
+    Please cite the following articles
+    .. [1] Neto Henriques, R., 2018. Advanced Methods for Diffusion MRI Data
+           Analysis and their Application to the Healthy Ageing Brain
+           (Doctoral thesis). https://doi.org/10.17863/CAM.29356
+    .. [2] Kellner E, Dhital B, Kiselev VG, Reisert M. Gibbs-ringing artifact
+           removal based on local subvoxel-shifts. Magn Reson Med. 2016
+           doi: 10.1002/mrm.26054.
+    """
+    if np.any(G0) is None or np.any(G1) is None:
+        G0, G1 = _weights(image.shape)
+
+    img_c1 = _gibbs_removal_1d(image, axis=1, n_points=n_points)
+    img_c0 = _gibbs_removal_1d(image, axis=0, n_points=n_points)
+
+    C1 = np.fft.fft2(img_c1)
+    C0 = np.fft.fft2(img_c0)
+    imagec = abs(np.fft.ifft2(np.fft.fftshift(C1)*G1 + np.fft.fftshift(C0)*G0))
+
+    return imagec
+
+
+def gibbs_removal(vol, slice_axis=2, n_points=3):
+    """ Suppresses gibbs ringing artefacts of images volumes.
+
+    Parameters
+    ----------
+    vol : array ([X, Y, Z]) or ([X, Y, Z, g])
+        Matrix containing one volume (3D) or multiple (4D) volumes of images.
+    slice_axis : int (0, 1, or 2)
```

Comment:
To be consistent with our framework, I think the user should format its data before calling your function.

Reply:
Indeed, I expect that the data is ordered in the following way ([X, Y, Z, g]); however that does not tell which axis contains the different 2D acquisitions. Normally, the last axis is used for different slices acquisition (i.e. you acquire slices with coordinates X, Y and align then in Z), but not sure how this is general for all datasets (e.g. I already saw a dataset in which different slices were aligned in X coordinates).

Reply:
Just to give a concrete example - imagine that you've acquired 60 slices with sampling matrix of 96x96. If you align image slices along the z axis, you will have a data matrix with shape (96, 96, 60). However, if user aligns the slices along the x-axis, your data matrix will have a shape of (60, 96, 96).  To use this code you will need to inform in which axis the slices are oriented because Gibbs suppression is done slice by slice.

Reply:
I understand your point but this is not coherent with all DIPY framework. the user will have to do the whole pipeline: denoise, unrig,  DTI  then tracking for example. So in your case, when data are flipped, it will work for your function but not the rest of the pipeline with strange to have 2 different logic. Give this opportunity to swap axis is misleading.  

It think it would be better for the user to flip the data at the beginning of the pipeline and then apply it to the whole process. 

Other opinions are welcomed

Reply:
@skoudoro - this is an optional parameter. If user don’t call it, function will assume that data is already aligned. So you can still used it if user flips the data before at the beginning of pipeline.

Also it is important to note that even if user decides to use this option to flip the data, Gibbs_removal will still return the data in its original shape. So I don’t see why it is not coherent with Dipy. 

Personally I don’t have any issue in removing this swap option. However, I really don’t see the harm of having this function with this feature. I think function in this way is more general.

Reply:
@skoudoro - this is an optional parameter. If user don’t call it, function will assume that data is already aligned. So you can still used it if user flips the data before at the beginning of pipeline.

Also it is important to note that even if user decides to use this option to flip the data, Gibbs_removal will still return the data in its original shape. So I don’t see why it is not coherent with Dipy. 

Personally I don’t have any issue in removing this swap option. However, I really don’t see the harm of having this function with this feature. I think function in this way is more general.

Reply:
Actually it is very dangerous to flip the data at the beginning of the pipeline. Because if user does it, we have to make sure that the affine matrix and gradient tables are also properly flipped. That is way, flipping and unflipping data should be done internally in this function.

Reply:
Actually it is very dangerous to flip the data at the beginning of the pipeline. Because if user does it, we have to make sure that the affine matrix and gradient tables are also properly flipped. That is way, flipping and unflipping data should be done internally in this function.

Reply:
> I expect that the data is ordered in the following way ([X, Y, Z, g]); however that does not tell which axis contains the different 2D acquisitions. Normally, the last axis is used for different slices acquisition (i.e. you acquire slices with coordinates X, Y and align then in Z), but not sure how this is general for all datasets (e.g. I already saw a dataset in which different slices were aligned in X coordinates)

ok, got it. you can let this option as it is. Thanks for the clarification!